src/main.rs

```diff
@@ -1311,78 +1311,99 @@ fn parse_manpage_at_path(
-//             sys.stdout.write("\r{0}\r".format(padded_progress_str))
-//             sys.stdout.flush()
-//
-//         # Maybe we want to skip this item
-//         skip = False
-//         if not WRITE_TO_STDOUT:
-//             # Compute the path that we would write to
-//             output_path = os.path.join(output_directory, output_file_name)
-//
-//         # Now skip if requested
-//         if skip:
-//             continue
-//
-//         try:
-//             if parse_manpage_at_path(manpage_path, output_directory):
-//                 successful_count += 1
-//         except IOError:
-//             diagnostic_indent = 0
-//             add_diagnostic('Cannot open ' + manpage_path)
-//         except (KeyboardInterrupt, SystemExit):
-//             raise
-//         except:
-//             add_diagnostic('Error parsing %s: %s' % (manpage_path, sys.exc_info()[0]), BRIEF_VERBOSE)
-//             flush_diagnostics(sys.stderr)
-//             traceback.print_exc(file=sys.stderr)
-//         flush_diagnostics(sys.stderr)
-//     print("") #Newline after loop
-//     add_diagnostic("Successfully parsed %d / %d pages" % (successful_count, total_count), BRIEF_VERBOSE)
-//     flush_diagnostics(sys.stderr)
-
 #[derive(Copy, Clone, Debug)]
 struct Progress(pub bool);
 
-// TODO Arg/output types?
+// Output: Result<cmd_name, error>
+// TODO: Result<cmd_name, CompletionsError>
 fn parse_and_output_man_pages(
-    _paths: impl Iterator<Item = PathBuf>,
-    _output_directory: PathBuf,
-    Progress(_show_progress): Progress,
+    paths: &mut [PathBuf],
+    output_directory: PathBuf,
+    Progress(show_progress): Progress,
+    deroff_only: bool,
+    write_to_stdout: bool,
 ) {
-    unimplemented!();
+    // Get the number of digits in num
+    let num_digits = |n| (n as f32).log10() as usize + 1;
+
+    let mut paths = paths;
+    paths.sort();
+
+    let total = paths.len();
+
+    let (mut successful_count, mut index) = (0, 0);
+    // TODO;
+    let mut cmd_name = "".to_owned();
+
+    let padding = num_digits(total);
+
+    let mut last_len: usize = 0;
+
+    if show_progress && !write_to_stdout {
+        println!(
+            "Parsing man pages and writing completions to {:?}",
+            output_directory
+        );
+    }
+
+    // As far as I can tell, we're fully supporting xz / lzma, so the checks here aren't needed
+
+    for manpage_path in paths {
+        index += 1;
+        // foo/bar/gcc.1.gz -> gcc.1.gz
+        // TODO: Expect
+        let man_file_name = manpage_path
+            .file_name()
+            .map(|fname| fname.to_string_lossy().to_mut().to_owned())
+            .unwrap_or_else(|| panic!("Failed to get manfile name from {:?}", manpage_path));
+
+        // gcc.1.gz -> gcc
+        cmd_name = man_file_name
+            .split('.')
+            .next()
+            .map(|cmd_name| cmd_name.to_owned())
+            .unwrap_or_else(|| panic!("Failed to get command name from {}", man_file_name));
```

Comment:
Seemed to be doing work twice by having 2 strings, `man_file_name` could be `&str` or borrowed without creating a new `String`.

Reply:
```rust
.map(|fname| fname.to_string_lossy())
```?

Reply:
I meant `to_owned()`, `...to_string_lossy` becomes `Cow` so it may not reallocate. I don't know what `to_mut` is for though. First, `to_owned`. Second, `cmd_name.to_owned()`.